### ISSUE TITLE
Research : Embedding Aware Attention

### ISSUE THREAD

**Issue by Optimox** [dreamquark-ai/tabnet#122]
# Main Problem

When training with large embedding dimensions, the mask size goes up.

One problem I see is that sparsemax does not know about which columns come from the same embedded columns, this could create something a bit difficult for the model to learn: 
- create embeddings that make sense
- mask embeddings without destroying them, in fact since sparsemax is sparse it's very unlikely that all the columns from a same embedding are used, so you lose the power of your embedding

# Proposed Solutions

It's an open problem but one way I see as promising is to create embedding aware attention.

 The idea would be to mask all dimensions from a same embedding the same way, either by using the mean or the max of the initial mask.

I implemented a first version here : #92 

If you feel like this is interesting and would like to contribute, please share your ideas in comments or open a PR!

**Comment by joseluismoreira** [comment]
Great issue! Reading the paper and the both original implementation and this one I made the same question about the embeddings from the same columns be used together.

**Comment by joseluismoreira** [comment]
I guess if we take the mean or the max we will lose the sparcity property... I am not sure, but maybe we can torch.stack the features instead of torch.cat them, and apply the sparse max in this new dimension.

**Comment by Optimox** [comment]
Hello @joseluismoreira,

There are two PR on this topic already:
- a very old one from me that was definitely not ideal : #92 
- a more recent one and more promising one : #217 

I guess the sparsity is not very important since you will only break the sparsity for embeddings from the same feature, so in the end you'll still be looking at only a few original features.

**Comment by joseluismoreira** [comment]
Thank you, @Optimox . The PR 217 seems what I look for. :+1:

**Comment by W-void** [comment]
After sparsemax,  only 1% feature is  nonzeros, is it normal？

**Comment by SeohuiPark** [comment]
I have a question about mask and "explain" method.

I checked on Tabmodel.py that there is an "explain" method of the model. I wonder how different masks are printed for each instance of the test data because the expand method does not train the model.

**Comment by Optimox** [comment]
@athewsey this has been around for way too long. I have been thinking of a more general way to deal with this and I created a PR here : #443

I'd be happy to have your thoughts on this!
